File: src/mastery/mastery-scoring.js
import { masteryScoringConfig } from "./scoring-config.js";

const stateRank = masteryScoringConfig.stateRank;
const confidenceLevelScore = masteryScoringConfig.confidenceLevelScore;
const readingScoreConfig = masteryScoringConfig.readingScore;
const trainingScoreConfig = masteryScoringConfig.trainingScore;
const stabilityScoreConfig = masteryScoringConfig.stabilityScore;
const conflictPenaltyConfig = masteryScoringConfig.conflictPenalty;
const targetReadinessConfig = masteryScoringConfig.targetReadiness;
const labelConfig = masteryScoringConfig.labels;

function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value));
}

function asArray(value) {
  return Array.isArray(value) ? value : [];
}

export function average(values = []) {
  if (!values.length) {
    return 0;
  }
  return Math.round(values.reduce((sum, value) => sum + value, 0) / values.length);
}

export function rankState(state = "不可判") {
  return stateRank[state] ?? stateRank.weak;
}

export function confidenceLevelToScore(level = "low") {
  return confidenceLevelScore[level] ?? confidenceLevelScore.low;
}

export function scoreToConfidenceLevel(score = 0) {
  if (score >= 0.75) {
    return "high";
  }
  if (score >= 0.45) {
    return "medium";
  }
  return "low";
}

export function buildMasteryLabel(score = 0, { hasEvidence = false, hasReading = false } = {}) {
  if (!hasEvidence && !hasReading) {
    return "未开始";
  }
  for (const entry of labelConfig.mastery) {
    if (score >= entry.min) {
      return entry.label;
    }
  }
  return "未开始";
}

export function buildTargetLabel(score = 0) {
  for (const entry of labelConfig.target) {
    if (score >= entry.min) {
      return entry.label;
    }
  }
  return "尚未建立证据";
}

export function calculateReadingScore(readingProgress = {}) {
  const progressPercentage = Number(readingProgress?.progressPercentage || 0);
  const dwellMs = Number(readingProgress?.dwellMs || 0);
  const completedReadCount = Number(readingProgress?.completedReadCount || 0);

  let score = 0;
  if (progressPercentage > 0) {
    score += readingScoreConfig.opened;
  }
  if (progressPercentage >= 25) {
    score += readingScoreConfig.progress25;
  }
  if (progressPercentage >= 50) {
    score += readingScoreConfig.progress50;
  }
  if (progressPercentage >= 75) {
    score += readingScoreConfig.progress75;
  }
  if (progressPercentage >= 90 && dwellMs >= 45_000) {
    score += readingScoreConfig.progress90Complete;
  }
  if (completedReadCount >= 2) {
    score += readingScoreConfig.secondFullReadBonus;
  }
  if (completedReadCount >= 3) {
    score += readingScoreConfig.thirdPlusFullReadBonus;
  }
  return clamp(score, 0, readingScoreConfig.max);
}

export function calculateTrainingScore(memoryItem = {}) {
  const evidenceCount = Number(memoryItem?.evidenceCount || 0);
  if (evidenceCount <= 0) {
    return 0;
  }

  const state = memoryItem?.state || "weak";
  const config = trainingScoreConfig[state] || trainingScoreConfig.weak;
  const score = config.base + Math.max(0, evidenceCount - 1) * config.repeat;
  return clamp(score, 0, Math.min(config.max, trainingScoreConfig.max));
}

export function calculateStabilityScore(memoryItem = {}) {
  const evidenceCount = Number(memoryItem?.evidenceCount || 0);
  if (evidenceCount <= 0) {
    return 0;
  }

  const state = memoryItem?.state || "不可判";
  const strongEvidenceCount = asArray(memoryItem?.recentStrongEvidence).length;
  if (state === "solid" && strongEvidenceCount >= 2) {
    return stabilityScoreConfig.repeatedSolid;
  }
  if (state === "solid" && strongEvidenceCount >= 1) {
    return stabilityScoreConfig.singleSolid;
  }
  if ((state === "partial" || state === "solid") && evidenceCount >= 2) {
    return stabilityScoreConfig.repeatedPartialOrAbove;
  }
  if (state === "partial" || state === "solid") {
    return stabilityScoreConfig.partialOrAbove;
  }
  return 0;
}

export function calculateConflictPenalty(memoryItem = {}) {
  const conflicts = asArray(memoryItem?.recentConflictingEvidence).length;
  if (!conflicts) {
    return 0;
  }

  let penalty = conflicts * conflictPenaltyConfig.perConflict;
  if ((memoryItem?.state || "不可判") === "weak") {
    penalty += conflictPenaltyConfig.weakStateExtra;
  }
  return clamp(penalty, 0, conflictPenaltyConfig.max);
}

export function calculateMasteryScore({ readingProgress = null, memoryItem = null } = {}) {
  const readingScore = calculateReadingScore(readingProgress || {});
  const trainingScore = calculateTrainingScore(memoryItem || {});
  const stabilityScore = calculateStabilityScore(memoryItem || {});
  const conflictPenalty = calculateConflictPenalty(memoryItem || {});

  return clamp(readingScore + trainingScore + stabilityScore - conflictPenalty, 0, 100);
}

export function calculateTargetReadinessScore(items = []) {
  if (!items.length) {
    return targetReadinessConfig.min;
  }

  const score = average(items.map((item) => Number(item?.masteryScore || 0)));
  return clamp(score, targetReadinessConfig.min, targetReadinessConfig.max);
}

File: ai-service/app/engine/mastery_scoring.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, List


CONFIG_PATH = Path(__file__).resolve().parents[3] / "contracts" / "mastery-scoring-v1.json"
CONFIG = json.loads(CONFIG_PATH.read_text(encoding="utf-8"))

STATE_RANK = CONFIG["stateRank"]
CONFIDENCE_LEVEL_SCORE = CONFIG["confidenceLevelScore"]
READING_SCORE = CONFIG["readingScore"]
TRAINING_SCORE = CONFIG["trainingScore"]
STABILITY_SCORE = CONFIG["stabilityScore"]
CONFLICT_PENALTY = CONFIG["conflictPenalty"]
TARGET_READINESS = CONFIG["targetReadiness"]
LABELS = CONFIG["labels"]


def clamp(value: float, minimum: float, maximum: float) -> float:
    return min(maximum, max(minimum, value))


def average(values: List[float]) -> int:
    if not values:
        return 0
    return round(sum(values) / len(values))


def rank_state(state: str = "不可判") -> int:
    return STATE_RANK.get(state, STATE_RANK["weak"])


def confidence_level_to_score(level: str = "low") -> float:
    return CONFIDENCE_LEVEL_SCORE.get(level, CONFIDENCE_LEVEL_SCORE["low"])


def score_to_confidence_level(score: float = 0.0) -> str:
    if score >= 0.75:
        return "high"
    if score >= 0.45:
        return "medium"
    return "low"


def build_mastery_label(score: float = 0.0, *, has_evidence: bool = False, has_reading: bool = False) -> str:
    if not has_evidence and not has_reading:
        return "未开始"
    for entry in LABELS["mastery"]:
        if score >= entry["min"]:
            return str(entry["label"])
    return "未开始"


def build_target_label(score: float = 0.0) -> str:
    for entry in LABELS["target"]:
        if score >= entry["min"]:
            return str(entry["label"])
    return "尚未建立证据"


def calculate_reading_score(reading_progress: Dict[str, Any] | None = None) -> int:
    reading_progress = reading_progress or {}
    progress_percentage = float(reading_progress.get("progressPercentage", 0) or 0)
    dwell_ms = float(reading_progress.get("dwellMs", 0) or 0)
    completed_read_count = int(reading_progress.get("completedReadCount", 0) or 0)

    score = 0
    if progress_percentage > 0:
        score += READING_SCORE["opened"]
    if progress_percentage >= 25:
        score += READING_SCORE["progress25"]
    if progress_percentage >= 50:
        score += READING_SCORE["progress50"]
    if progress_percentage >= 75:
        score += READING_SCORE["progress75"]
    if progress_percentage >= 90 and dwell_ms >= 45_000:
        score += READING_SCORE["progress90Complete"]
    if completed_read_count >= 2:
        score += READING_SCORE["secondFullReadBonus"]
    if completed_read_count >= 3:
        score += READING_SCORE["thirdPlusFullReadBonus"]
    return int(clamp(score, 0, READING_SCORE["max"]))


def calculate_training_score(memory_item: Dict[str, Any] | None = None) -> int:
    memory_item = memory_item or {}
    evidence_count = int(memory_item.get("evidenceCount", 0) or 0)
    if evidence_count <= 0:
        return 0

    state = str(memory_item.get("state", "weak"))
    config = TRAINING_SCORE.get(state, TRAINING_SCORE["weak"])
    score = config["base"] + max(0, evidence_count - 1) * config["repeat"]
    return int(clamp(score, 0, min(config["max"], TRAINING_SCORE["max"])))


def calculate_stability_score(memory_item: Dict[str, Any] | None = None) -> int:
    memory_item = memory_item or {}
    evidence_count = int(memory_item.get("evidenceCount", 0) or 0)
    if evidence_count <= 0:
        return 0

    state = str(memory_item.get("state", "不可判"))
    strong_evidence_count = len(memory_item.get("recentStrongEvidence") or [])
    if state == "solid" and strong_evidence_count >= 2:
        return int(STABILITY_SCORE["repeatedSolid"])
    if state == "solid" and strong_evidence_count >= 1:
        return int(STABILITY_SCORE["singleSolid"])
    if state in {"partial", "solid"} and evidence_count >= 2:
        return int(STABILITY_SCORE["repeatedPartialOrAbove"])
    if state in {"partial", "solid"}:
        return int(STABILITY_SCORE["partialOrAbove"])
    return 0


def calculate_conflict_penalty(memory_item: Dict[str, Any] | None = None) -> int:
    memory_item = memory_item or {}
    conflicts = len(memory_item.get("recentConflictingEvidence") or [])
    if conflicts <= 0:
        return 0

    penalty = conflicts * CONFLICT_PENALTY["perConflict"]
    if str(memory_item.get("state", "不可判")) == "weak":
        penalty += CONFLICT_PENALTY["weakStateExtra"]
    return int(clamp(penalty, 0, CONFLICT_PENALTY["max"]))


def calculate_mastery_score(*, reading_progress: Dict[str, Any] | None = None, memory_item: Dict[str, Any] | None = None) -> int:
    return int(
        clamp(
            calculate_reading_score(reading_progress)
            + calculate_training_score(memory_item)
            + calculate_stability_score(memory_item)
            - calculate_conflict_penalty(memory_item),
            0,
            100,
        )
    )


def calculate_target_readiness_score(items: List[Dict[str, Any]]) -> int:
    if not items:
        return int(TARGET_READINESS["min"])
    score = average([float(item.get("masteryScore", 0) or 0) for item in items])
    return int(clamp(score, TARGET_READINESS["min"], TARGET_READINESS["max"]))
